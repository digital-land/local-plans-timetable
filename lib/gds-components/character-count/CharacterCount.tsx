/**
 * GDS component: https://design-system.service.gov.uk/components/character-count/
 */

export interface CharacterCountProps {
  characterLimit: number;
  id?: string;
  numberOfCharacters: number;
}

const getRemainingCharactersMessage = (
  numberOfCharacters: number,
  characterLimit: number
) => {
  const remainingCharacters = characterLimit - numberOfCharacters;
  const absoluteRemainingCharacters = Math.abs(remainingCharacters);

  return `You have ${absoluteRemainingCharacters} character${
    absoluteRemainingCharacters === 1 ? "" : "s"
  } ${remainingCharacters < 0 ? "too many" : "remaining"}`;
};

export const CharacterCount = ({
  characterLimit,
  id,
  numberOfCharacters,
}: CharacterCountProps) => (
  <div id={id} className="govuk-hint">
    {getRemainingCharactersMessage(numberOfCharacters, characterLimit)}
  </div>
);
