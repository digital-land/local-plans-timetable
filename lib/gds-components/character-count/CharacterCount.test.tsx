import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";

import { CharacterCount, CharacterCountProps } from "./CharacterCount";

interface TestCase extends Omit<CharacterCountProps, "id"> {
  expectedMessage: string;
}

const testCases: TestCase[] = [
  {
    characterLimit: 10,
    numberOfCharacters: 5,
    expectedMessage: "You have 5 characters remaining",
  },
  {
    characterLimit: 10,
    numberOfCharacters: 9,
    expectedMessage: "You have 1 character remaining",
  },
  {
    characterLimit: 10,
    numberOfCharacters: 10,
    expectedMessage: "You have 0 characters remaining",
  },
  {
    characterLimit: 10,
    numberOfCharacters: 11,
    expectedMessage: "You have 1 character too many",
  },
  {
    characterLimit: 10,
    numberOfCharacters: 15,
    expectedMessage: "You have 5 characters too many",
  },
];

test.each(testCases)(
  "renders the correct message for numberOfCharacters === $numberOfCharacters, characterLimit === $characterLimit",
  ({ characterLimit, numberOfCharacters, expectedMessage }) => {
    const { asFragment } = render(
      <CharacterCount
        characterLimit={characterLimit}
        numberOfCharacters={numberOfCharacters}
      />
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.queryByText(expectedMessage)).toBeInTheDocument();
  }
);
