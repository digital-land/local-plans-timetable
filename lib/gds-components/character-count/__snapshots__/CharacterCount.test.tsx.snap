// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct message for numberOfCharacters === 5, characterLimit === 10 1`] = `
<DocumentFragment>
  <div
    class="govuk-hint"
  >
    You have 5 characters remaining
  </div>
</DocumentFragment>
`;

exports[`renders the correct message for numberOfCharacters === 9, characterLimit === 10 1`] = `
<DocumentFragment>
  <div
    class="govuk-hint"
  >
    You have 1 character remaining
  </div>
</DocumentFragment>
`;

exports[`renders the correct message for numberOfCharacters === 10, characterLimit === 10 1`] = `
<DocumentFragment>
  <div
    class="govuk-hint"
  >
    You have 0 characters remaining
  </div>
</DocumentFragment>
`;

exports[`renders the correct message for numberOfCharacters === 11, characterLimit === 10 1`] = `
<DocumentFragment>
  <div
    class="govuk-hint"
  >
    You have 1 character too many
  </div>
</DocumentFragment>
`;

exports[`renders the correct message for numberOfCharacters === 15, characterLimit === 10 1`] = `
<DocumentFragment>
  <div
    class="govuk-hint"
  >
    You have 5 characters too many
  </div>
</DocumentFragment>
`;
